src/jit/codegenxarch.cpp

```diff
@@ -1447,13 +1447,13 @@ void CodeGen::genCodeForBinary(GenTree* treeNode)
         && src->isContainedIntOrIImmed() 
         && !treeNode->gtOverflowEx())
     {
-        if (src->gtIntConCommon.IconValue() == 1)
+        if (src->IsIntegralOne())
```

Comment:
The if-check preceding this one has already established it is GT_CNS_INT.  But IsIntegralOne() is once again checking whether gtOper is GT_CNS_INT or GT_CNS_LNG, which is unnecessary.  To be precise, above and on line 1456 we should use IsIntOne() or IsIntMinusOne().  Even then we incur additional gtOper check inside IsIntOne() and IsIntMinusOne(). I see this excessive check in many places due to the replacement of original condition by IsIntOne/IsIntegralOne/IsIntMinusOne/IsIntegralMinusOne().

The question I have is, Is native compiler clever enough to optimize out additional gtOper check done by IsIntOne()/IsIntMinuOne() or is it best to avoid it in the first place? Note that this we need to see not just above but in all the places where usage of IsIntXXX()/IsIntegralXXX() has resulted in an additional gtOper check.

I am thinking whether these methods be defined on GenTreeIntConCommon and GenTreeDblCon instead of on GenTree?  Another advantage is that we can drop the "type" from these method names when defined on GenTreeIntCon, GenTreeLngCon and GenTreeDblCon

e.g
/ / actual methods
bool GenTreeIntCon::IsOne()
bool GenTreeIntCon::IsZero()
bool GenTreeIntCon::IsMinusOne()
bool GenTreeLngCon::IsOne()
bool GenTreeLngCon::IsZero()
bool GenTreeLngCon::IsMinusOne()
bool GenTreeDblCon::IsZero()

In case if we want to have similar methods defined on GenTree, we can do so and such a method implementation based on gtOper will cast GenTree node to specific *Con node and call the corresponding method on it.  For example GenTree:IsZero() could be implemented as follows assuming GenTreeIntCon::IsZero(), GenTreeLngCon::IsZero and GenTreeDblCon::IsZero() are available

```
bool GenTree::IsZero() 
{
   if (OperGet() == GT_CNS_INT)
   {
       this->AsIntCon()->IsZero();
   }
  else if (OperGet() == GT_CNS_LNG)
   {
        this->AsLngCon()->IsZero();
   }
  else if (OperGet() == GT_CNS_DBL) 
  {
       this->AsDblCon()->IsZero();
  }
  return false;
}

```
With the above suggestion, if-checks on lines 1450 and 1456 would become as follows

```
if (src->AsIntCon()->IsOne()) 
if (src->AsIntCon()->IsMinusOne())
```
Note: Ideally if we used virtual methods it would have allowed us to write
`if (src->IsOne()) or if (src->IsMinusOne())`.  Since we don't use virtual methods in Gentree nodes for efficiency, either 1) we have to check gtOper of the node, cast the node statically to right type and call the method IsOne()/IsMinusOne() on precise Gentree type node (**OR**) 2) GenTree::IsOne()/IsMinusOne() implementation should check gtOper and return the right result.   If there are many places where we already check gtOper and just further need to check whether const value is 0, 1 or -1, option 1 would avoid unnecessary additional gtOper check.  Option 2 would be useful in places where we need to do both gtOper check and const value check together.